src/number_theory/legendre_symbol/quadratic_char.lean

```diff
@@ -0,0 +1,339 @@
+end
+
+/-- A unit `a` of a finite field `F` of odd characteristic is a square
+if and only if `a ^ (#F / 2) = 1`. -/
+lemma unit_is_sqare_iff (hF : ring_char F ≠ 2) (a : Fˣ) :
+  is_square a ↔ a ^ (fintype.card F / 2) = 1 :=
+begin
+  classical,
+  obtain ⟨g, hg⟩ := is_cyclic.exists_generator Fˣ,
+  obtain ⟨n, hn⟩ : a ∈ submonoid.powers g, { rw mem_powers_iff_mem_zpowers, apply hg },
+  have hodd := nat.two_mul_odd_div_two (finite_field.odd_card_of_char_ne_two hF),
+  split,
+  { rintro ⟨y, rfl⟩,
+    rw [← pow_two, ← pow_mul, hodd],
+    apply_fun (@coe Fˣ F _),
+    push_cast,
+    exact finite_field.pow_card_sub_one_eq_one (y : F) (units.ne_zero y),
+    exact units.ext, },
+  { subst a, assume h,
+    have key : 2 * (fintype.card F / 2) ∣ n * (fintype.card F / 2),
+    { rw [← pow_mul] at h,
+      rw [hodd, ← fintype.card_units, ← order_of_eq_card_of_forall_mem_zpowers hg],
+      apply order_of_dvd_of_pow_eq_one h },
+    have : 0 < fintype.card F / 2 := nat.div_pos fintype.one_lt_card (by norm_num),
+    obtain ⟨m, rfl⟩ := nat.dvd_of_mul_dvd_mul_right this key,
+    refine ⟨g ^ m, _⟩,
+    rw [mul_comm, pow_mul, pow_two], },
+end
+
+/-- A non-zero `a : F` is a square if and only if `a ^ (#F / 2) = 1`. -/
+lemma is_square_iff (hF : ring_char F ≠ 2) {a : F} (ha : a ≠ 0) :
+  is_square a ↔ a ^ (fintype.card F / 2) = 1 :=
+begin
+  apply (iff_congr _ (by simp [units.ext_iff])).mp
+        (finite_field.unit_is_sqare_iff hF (units.mk0 a ha)),
+  simp only [is_square, units.ext_iff, units.coe_mk0, units.coe_mul],
+  split, { rintro ⟨y, hy⟩, exact ⟨y, hy⟩ },
+  { rintro ⟨y, rfl⟩,
+    have hy : y ≠ 0, { rintro rfl, simpa [zero_pow] using ha, },
+    refine ⟨units.mk0 y hy, _⟩, simp, }
+end
+
+/-- In a finite field of odd characteristic, not every element is a square. -/
+lemma exists_nonsquare (hF : ring_char F ≠ 2) : ∃ (a : F), ¬ is_square a :=
+begin
+  -- idea: the squaring map on `F` is not injetive, hence not surjective
+  let sq : F → F := λ x, x^2,
+  have h : ¬ function.injective sq,
+  begin
+    simp only [function.injective, not_forall, exists_prop],
+    use [-1, 1],
+    split,
+    { simp only [sq, one_pow, neg_one_sq], },
+    { exact finite_field.neg_one_ne_one_of_char_ne_two hF, },
+  end,
+  have h₁ := mt (fintype.injective_iff_surjective.mpr) h, -- sq not surjective
+  push_neg at h₁,
+  cases h₁ with a h₁,
+  use a,
+  simp only [is_square, sq, not_exists, ne.def] at h₁ ⊢,
+  intros b hb,
+  rw ← pow_two at hb,
+  exact (h₁ b hb.symm),
+end
+
+end finite_field
+
+end general
+
+namespace char
+
+/-!
+### Definition of the quadratic character
+
+We define the quadratic character of a finite field `F` with values in ℤ.
+-/
+
+section define
+
+/-- Define the quadratic character of a finite field `F` with values in ℤ.
+It takes the value zero at zero; for non-zero argument `a : F`, it is `1`
+if `a` is a square, otherwise it is `-1`.
+-/
+def quadratic_char (F : Type*) [field F] [fintype F] [decidable_eq F] (a : F) : ℤ :=
```

Comment:
There is now a type called `sign_type` (was recently added). Would it make sense to take that as the target of this defn? Also, this defn doesn't require `field` or `fintype`, right? (I agree that it maybe doesn't make sense for random `monoid_with_zero`s, but it doesn't hurt either to make the definition general.)

Reply:
I'm not sure that using `sign_type` would offer advantages. I would like to be able to do computations like `quadratic_char_sum_zero` or Gauss sums with the values, for which one would have to cast the `sign_type` values to integers (or elements of some other ring).

I'll reduce the requirements to `[has_mul F]` (and use a different name than `F`); this is what `is_square` needs.

Reply:
With `[decidable_pred (is_square : F \to Prop)]` in addition to `[monoid_with_zero F] [decidable_eq F]` it seems to work.